### ISSUE TITLE
Slow queries till Nextcloud 30.01

### ISSUE THREAD

**Issue by anoriel** [nextcloud/groupfolders#3376]
### Steps to reproduce
1. migrate from 29.0.x to 30.0.1
2. display files page
3. show processes list in mariadb

### Expected behaviour
No delay as previous version

### Actual behaviour
The page is loading for a long time, nearly 20sec.
It seems to be caused by slow queries in mariadb (example below)

### Server configuration

**Operating system: Ubuntu 22.04

**Web server: nginx

**Database: mariadb 10.6.18

**PHP version: 8.3

**Nextcloud version: 30.0.1

**Group folders version: 18.0.3

**Are you using external storage, if yes which one: no

**Are you using encryption: no

**Are you using an external user-backend, if yes which one: LDAP

### Client configuration
**Browser: all (Chrome, firefox, safari...)

**Operating system: all (Mac, Windows, Linux)

### Logs

example of slow query:
`SELECT f.folder_id, f.mount_point, f.quota, f.acl, c.fileid, c.storage, c.path, c.name, c.mimetype, c.mimepart, c.size, c.mtime, c.storage_mtime, c.etag, c.encrypted, c.parent, a.permissions AS group_permissions, c.permissions AS permissions
FROM oc_group_folders f
INNER JOIN oc_group_folders_groups a ON f.folder_id = a.folder_id
LEFT JOIN oc_filecache c ON (c.name = CONCAT(f.folder_id, '')) AND (c.parent = '12345')
LEFT JOIN oc_circles_membership w_j ON (w_j.single_id = 'abcdefghijklmnopqrstuvwxyz12345') AND (w_j.circle_id = a.circle_id)
LEFT JOIN oc_circles_circle w_j_k ON w_j.circle_id = w_j_k.unique_id
WHERE w_j.level >= 1;`

**Comment by anoriel** [comment]
I forgot to mention that all indexes are ok, no repair command is waiting, no warning in nextcloud/admin page...

**Comment by provokateurin** [comment]
I suppose you already had the Circles/Teams app enabled before?
How many of those slow queries are you seeing concurrently when you do a single reload of a Groupfolder?
Are all the queries the same or are there different ones as well?

I'm not aware of any changes we did to the Database between 29 and 30, so I don't know if this is actually Groupfolders related or Server or your infrastructure. Can you check if you encounter any other performance problems after you upgraded to 30 (somewhere else, not in Groupfolders)?

**Comment by anoriel** [comment]
Yes Circles/Teams was enable before
Nothing has been added or installed between 29 and 30, we only upgraded the core NC.
We have ~40 simultaneous queries but we got ~50 users working daily so it's not easy to isolate one user's queries...
Each query is taking from 2,5 to 6 seconds to execute so the main website is slowed down by this...

**Comment by anoriel** [comment]
All other apps seem to work perfectly (talk, calendar,...)
Only Files tab is affected

**Comment by anoriel** [comment]
If I remove the left join to filecache table, the query is executed instantanesously... 
As this table is about 13Gb, it's not a surprise...

If I change the query to something like this, the answer is also executed quickely:

`SELECT a.*, c.permissions AS permissions, c.fileid,  c.storage,  c.path,  c.name, c.mimetype, c.mimepart, c.size, c.mtime, c.storage_mtime, c.etag, c.encrypted, c.parent 
FROM (
  SELECT f.folder_id, 
    f.mount_point, 
    f.quota, 
    f.acl, 
    a.permissions AS group_permissions
  FROM oc_group_folders f
  INNER JOIN oc_group_folders_groups a ON f.folder_id = a.folder_id
  LEFT JOIN oc_circles_membership w_j ON (w_j.single_id = 'abcdefghijklmnopqrstuvwxyz12345') AND (w_j.circle_id = a.circle_id)
  LEFT JOIN oc_circles_circle w_j_k ON w_j.circle_id = w_j_k.unique_id
  WHERE w_j.level >= 1
) a
LEFT JOIN oc_filecache c ON (c.name = CONCAT(a.folder_id, '')) AND (c.parent = '12345')
;`

Do you think you can update the app to change the left join to filecache table query?

**Comment by enbata** [comment]
Hi:

we are facing also slow queries after upgrade to 30.0.4 with this query:

`SELECT f.folder_id, mount_point, quota, acl, c.fileid, c.storage, c.path, c.name, c.mimetype, c.mimepart, c.size, c.mtime, c.storage_mtime, c.etag, c.encrypted, c.parent, a.permissions AS group_permissions, c.permissions AS permissions FROM oc_group_folders f INNER JOIN oc_group_folders_groups a ON f.folder_id = a.folder_id LEFT JOIN oc_filecache c ON (c.name = CONCAT(f.folder_id, '')) AND (c.parent = '183') AND (c.storage = '1') WHERE a.group_id IN ('X', 'Y', 'Z');`

The MySql ANALYZE result:

![Image](https://github.com/user-attachments/assets/a6c427a2-bf83-4f30-b1b7-21065f01a5de)

It tooks more than 1 second to resolve, it seems that is not using the indexes.

If we delete `c.name = CONCAT(f.folder_id, '')` condition or change `a.group_id IN ('X', 'Y', 'Z')` to only one `a.group_id IN ('X', 'Y', 'Z')` it executes quickly.

![Image](https://github.com/user-attachments/assets/2919c2a1-ec6a-410c-977d-786cf7f68cd0)

We think is related with @anoriel's problem. We see that the version 19.0.0 is on *alpha*, it will be changed/fixed with that version?

Thanks,

Manex

**Comment by CedBadarou** [comment]
I'm facing the same issue with Team folders 19.1.0. I still need more than 10 secs to load any groupfolder.
In my case I wonder if it's not due to the large amount of mounts I have (over 500000).
It could be related to [1506](https://github.com/nextcloud/groupfolders/issues/1506).

**Comment by provokateurin** [comment]
That's indeed quite a lot and I suppose there are some queries that can be removed or optimized. Could you perhaps check your slow queries log and post them here? Any data on what queries are actually running and how often and how long they take would be really valuable!

**Comment by CedBadarou** [comment]
Yes I will do that asap. Though I don't know if there are slow queries or if there are too many queries...

**Comment by CedBadarou** [comment]
So I have something, I don't know if it will help, I'll try to do better tomorrow.

When opening a team folder, I can see this :

- A query ```SELECT "path" FROM "oc_filecache" WHERE ("storage" = $1) AND ("path_hash" IN (1001 parameters))``` and another one with 155 parameters
- 1177 `INSERT INTO "oc_file_locks"` queries
- 1159 ```SELECT "fileid" FROM "oc_filecache" WHERE ("storage" = $1) AND ("path_hash" = $2)```

I'm not a good sql debugger so I'm not sure about the durations I'm seeing (is `duration: 6.106 ms` 6 seconds or 6 ms ?), I will tell you more about that tomorrow. But I think the problem lies more in the amount of queries than in their durations.

**Comment by CedBadarou** [comment]
Ok so setting database `log_min_duration_statement` to 500, I can see 39 `UPDATE "oc_file_locks" SET "lock" = "lock" - '1' WHERE ("key" IN (...)) AND ("lock" > 0)` queries, which all took around 1 sec.

**Comment by CedBadarou** [comment]
So I disabled filelocking, and now I don't see any query taking more than 200ms when I load the files page. But it took over 5 seconds anyway, which is way better than 20, but still longer than it used to be.
